File: packages/datadog/python/zerithdb_datadog/formatter.py
from __future__ import annotations
import datetime, os, socket
from typing import Any, Literal, TypedDict

LogLevel = Literal["debug", "info", "warn", "error", "critical"]

class ZerithLogEntry(TypedDict, total=False):
    level:     LogLevel
    message:   str
    timestamp: datetime.datetime
    context:   dict[str, Any]

class DatadogPluginOptions(TypedDict, total=False):
    service: str
    host:    str
    source:  str
    tags:    list[str]
    env:     str
    version: str

# Datadog status values per level
_LEVEL_MAP: dict[str, str] = {
    "debug":    "debug",
    "info":     "info",
    "warn":     "warning",
    "error":    "error",
    "critical": "critical",
}

def format_for_datadog(
    entry: ZerithLogEntry,
    opts:  DatadogPluginOptions,
) -> dict[str, Any]:
    """Return a dict formatted for Datadog log ingestion."""
    now = entry.get("timestamp") or datetime.datetime.utcnow()
    tags: list[str] = []
    if env := opts.get("env"):
        tags.append(f"env:{env}")
    if ver := opts.get("version"):
        tags.append(f"version:{ver}")
    tags.extend(opts.get("tags", []))

    payload: dict[str, Any] = {
        "message":   entry["message"],
        "status":    _LEVEL_MAP.get(entry.get("level", "info"), "info"),
        "service":   opts["service"],
        "ddsource":  opts.get("source", "zerithdb"),
        "ddtags":    ",".join(tags),
        "host":      opts.get("host") or socket.gethostname(),
        "timestamp": now.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z",
        "logger":    {"name": "zerithdb-datadog"},
    }
    payload.update(entry.get("context", {}))  # spread extra fields
    return payload

File: packages/datadog/src/formatter.ts
import { ZerithLogEntry, DatadogLogPayload, DatadogPluginOptions, LogLevel } from "./types";

/** Datadog status string per log level */
const LEVEL_MAP: Record<LogLevel, string> = {
  debug:    "debug",
  info:     "info",
  warn:     "warning",
  error:    "error",
  critical: "critical",
};

export function formatForDatadog(
  entry: ZerithLogEntry,
  opts: DatadogPluginOptions
): DatadogLogPayload {
  const tags: string[] = [
    opts.env     ? `env:${opts.env}`         : "",
    opts.version ? `version:${opts.version}` : "",
    ...(opts.tags ?? []),
  ].filter(Boolean);

  return {
    message:   entry.message,
    status:    LEVEL_MAP[entry.level] ?? "info",
    service:   opts.service,
    ddsource:  opts.source ?? "zerithdb",
    ddtags:    tags.join(","),
    host: opts.host ?? (typeof process !== "undefined" ? process.env.HOSTNAME ?? "unknown" : "unknown"),
    timestamp: (entry.timestamp ?? new Date()).toISOString(),
    logger:    { name: "zerithdb-datadog" },
    ...entry.context,  // spread extra fields (peerId, appId, etc.)
  };
}

/** Batch-format multiple entries */
export function formatBatch(
  entries: ZerithLogEntry[],
  opts: DatadogPluginOptions
): DatadogLogPayload[] {
  return entries.map((e) => formatForDatadog(e, opts));
}